### PR TITLE
chore: Move boto3-stubs dependency from dev.txt to base.txt

### PR DESCRIPTION
**Pull request by hnnasit**
#### Why is this change necessary?
`boto3-stubs` is a dev dependency and to use the boto3 objects for type checking, they need to be imported conditionally. If the import is not conditional, it can break production installers as the testing uses dev-dependencies and this can get missed.

#### How does it address the issue?
This PR moves `boto3-stubs` to base.txt for making the boto3 type checks more sustainable. I built a mac binary with the new changes and tested the functionality for `remote invoke` and `sam deploy` companion stack and it works as expected.

#### What side effects does this change have?
N/A

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [x] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Review by mndeveci — APPROVED** [review]
LGTM, can we confirm generated artifacts with E2E tests that it doesn't have any side effects?

**Comment by hnnasit** [comment]
> LGTM, can we confirm generated artifacts with E2E tests that it doesn't have any side effects?

I did some testing around `remote invoke` and `sam deploy` commands after generating an artifact with the new changes. I can run the E2E integ tests to verify as well before merging.

File: requirements/dev.txt [hunk 1]
```diff
@@ -8,7 +8,6 @@ pytest-cov==4.1.0
 # mypy adds new rules in new minor versions, which could cause our PR check to fail
 # here we fix its version and upgrade it manually in the future
 mypy==1.3.0
-boto3-stubs[apigateway,cloudformation,ecr,iam,lambda,s3,schemas,secretsmanager,signer,stepfunctions,sts,xray]==1.28.7
```
Comment (hawflau) [inline requirements/dev.txt:11]: not an immediate issue but we should think about it - I wonder if we should move all typing stubs to base.txt, considering that the same could happen for any typing stubs we are using. That also yield another questions:
- How do we test if we've included all dependencies used for type hints into base.txt?
- Should we consider striping all type hints as part of release process after passing type checks?

**Review by hawflau — APPROVED** [review]

File: requirements/dev.txt (hunk 1 above)
Comment (hnnasit) [reply requirements/dev.txt:11]: I think that's a good callout and both are valid questions worth thinking.

**Pull request merged**